Convert ActionCable javascript to ES2015 modules with a modern build environment

rmacklin opened the pull request:
### Summary

This pull request migrates ActionCable's javascript source code and tooling to mirror that of the more modern ActiveStorage package.

The goals are:
1. Create consistency in the javascript environments for ActionCable and ActiveStorage
   - The aim is to converge on how each package handles the file structure, linting, build pipeline, etc.
2. Migrate ActionCable from CoffeeScript to ES2015 and from sprockets `//= require` directives to ES2015 module syntax
   - This is consistent with the more modern style of ActiveStorage's javascript. Most of the front end community has moved away from CoffeeScript and embraced ES2015+, so it seems like a good idea to modernize ActionCable's front end using the same technologies as ActiveStorage.
3. Update the ActionCable npm package to ship untranspiled ES2015 source code in addition to the heavier compiled bundle
   - This enables ActionCable users to ship smaller javascript bundles to their visitors with modern browsers. We added this enhancement to ActiveStorage in https://github.com/rails/rails/pull/31880.

### Other Information

As in https://github.com/rails/rails/pull/31880, I've created a sample repository to demonstrate the benefits of goal (3):
https://github.com/rmacklin/actioncable-es2015-build-example

In that example, the bundle shrinks by 2.5K (23%) when you simply change the actioncable import to point to the untranspiled src, e.g. by [setting up an alias](https://github.com/rmacklin/actioncable-es2015-build-example/commit/278154577d85e9e8f3231b1d268d12f8a3fd49af):

```
resolve: {
  alias: {
    actioncable: 'actioncable/src'
  }
}
```

If you go a step further, by [using a fine-grained import to pull in only what your app uses](https://github.com/rmacklin/actioncable-es2015-build-example/commit/d494d6bf160042ca218bf3bd7aae6c1929492e0b), e.g.:
```diff
diff --git a/app/scripts/main.js b/app/scripts/main.js
index 17bc031..1a2b2e0 100644
--- a/app/scripts/main.js
+++ b/app/scripts/main.js
@@ -1,6 +1,6 @@
-import ActionCable from 'actioncable';
+import {createConsumer} from 'actioncable/action_cable';

-let cable = ActionCable.createConsumer('wss://cable.example.com');
+let cable = createConsumer('wss://cable.example.com');

 cable.subscriptions.create('AppearanceChannel', {
```

then the bundle shrinks by 3.2K (30%).

In addition to allowing smaller bundles for those who ship untranspiled code to modern browsers, including the source code in the published package can be useful in other ways:

1. Users can import individual modules rather than the whole library
2. As a result of (1), users can also monkey patch parts of ActionCable by importing the relevant module, modifying the exported object, and then importing the rest of ActionCable (which would then use the patched object).

### PR Breakdown

Given the scope of the changes, this is not a small PR. I did my best to include thorough commit messages. Here are the high level chunks of how the PR breaks down:
1. Convert regular CoffeeScript files to ES2015 (but not ES2015 modules yet) using [decaffeinate](https://github.com/decaffeinate/decaffeinate) https://github.com/rails/rails/compare/61f5ba7156a8050b48380765ceaac716a4e8296b~...0398e595f9823a8c40224be2e7406bca66bb624e
2. Clean up the converted code https://github.com/rails/rails/compare/cbe65e042877976c2e1b52bbe6ae74f850ce15c6~...8ce698740e27acd8769574d652f65485ab4dc6cf
3. Convert `action_cable.coffee.erb` and do more cleanup https://github.com/rails/rails/compare/8f23df109b25b5d33da09246e806fbfa9e765dc8~...bba49e0c172fd37401bf05888e79dc18e75b06e1
4. Modernize ActionCable tooling/environment to mirror ActiveStorage's setup https://github.com/rails/rails/compare/3ae4ebfbfdd407ce09990bf5daf306078937876c~...575ce04234bd3e5e443c4475efef55f86c4e462d
5. Convert ActionCable to ES2015 modules https://github.com/rails/rails/compare/8936336a7e957e51b9b0ff185c8885b395a537ab~...32e50483df697b1ad5dcc33aa5914767ed8f454a
6. Final cleanup and finishing touches https://github.com/rails/rails/compare/1866d0b77c90fbb18db087c4c09cd08e09c595d2~...83fbeee2c864df282feb1b13487900632d194139
7. Include source code in published npm package https://github.com/rails/rails/commit/13b429a0cd01775bea375882521294e6eab3605a

Unfortunately, splitting off separate PRs for each chunk doesn't work because you have to get all the way to ES2015 modules to leverage the rollup build pipeline. (The build process is broken in the intermediate stages, and there's not an easy way to avoid that given the limitations of sprockets.)

---

cc @javan - I'd love to hear your thoughts on this, and I have some additional questions and ideas that we could discuss to improve this.

---

Edit: Goal (3) has been split out to a new PR:
- https://github.com/rails/rails/pull/34370

rmacklin committed:
Move actioncable javascript to app/javascript and change .coffee -> .js

- Rename action_cable/*.coffee -> *.js

- Move app/assets/javascripts/* -> app/javascript/*

- Rename action_cable.js.erb -> action_cable/index.js.erb

Renaming the extension to .js is in preparation for converting these
files from coffeescript to ES2015.

Moving the files to app/javascript and putting the entry point in
index.js.erb changes the structure of ActionCable's javascript to match
the structure of ActiveStorage's javascript.

(We are doing the file moving and renaming in a separate commit to
ensure that the git history of the files will be preserved - i.e. git
will track these as file renames rather than unrelated file
additions/deletions. In particular, git blame will still trace back to
the original authorship.)

rmacklin committed:
Run decaffeinate on action_cable/*.js

Using [decaffeinate], we have converted these files from coffeescript
syntax to ES2015 syntax. Decaffeinate is very conservative in the
conversion process to ensure exact coffeescript semantics are preserved.
Most of the time, it's safe to clean up the code, and decaffeinate has
left suggestions regarding potential cleanups we can take. I'll tackle
those cleanups separately.

After running decaffeinate, I ran:
```
eslint --fix app/javascript
```
using the eslint configuration from ActiveStorage to automatically
correct lint violations in the decaffeinated output. This removed 189
extra semicolons and changed one instance of single quotes to double
quotes.

Note: decaffeinate and eslint can't parse ERB syntax. So I worked around
that by temporarily quoting the ERB:
```diff
 @ActionCable =
-  INTERNAL: <%= ActionCable::INTERNAL.to_json %>
+  INTERNAL: "<%= ActionCable::INTERNAL.to_json %>"
   WebSocket: window.WebSocket
   logger: window.console
```
and then removing those quotes after running decaffeinate and eslint.

[decaffeinate]: https://github.com/decaffeinate/decaffeinate

rmacklin committed:
Refactor decaffeinate output to more natural/idiomatic javascript

- Remove unnecessary Array.from usages from subscriptions.js

  These were all Arrays before, so Array.from is a no-op

- Remove unnecessary IIFEs from subscriptions.js

- Manually decaffeinate sample ActionCable code in comments

  Here the coffeescript -> ES2015 conversion was done by hand rather than
  using decaffeinate, because these code samples were simple enough.

- Refactor ActionCable.Subscription to avoid initClass

- Refactor ActionCable.Subscription to use ES2015 default parameters

- Refactor ActionCable.ConnectionMonitor to avoid initClass

- Refactor ActionCable.ConnectionMonitor to use shorter variations of null checks

- Remove unnecessary code created because of implicit returns in ConnectionMonitor

  This removes the `return` statements that were returning the value of
  console.log and those from private methods whose return value was not
  being used.

- Refactor ActionCable.Connection to avoid initClass

- Refactor Connection#isProtocolSupported and #isState

  This addresses these three decaffeinate cleanup suggestions:
  - DS101: Remove unnecessary use of Array.from
  - DS104: Avoid inline assignments
  - DS204: Change includes calls to have a more natural evaluation order

  It also removes the use of Array.prototype.includes, which means we
  don't have to worry about providing a polyfill or requiring that end
  users provide one.

- Refactor ActionCable.Connection to use ES2015 default parameters

- Refactor ActionCable.Connection to use shorter variations of null checks

- Remove return statements that return the value of console.log() in ActionCable.Connection

- Simplify complex destructure assignment in connection.js

  decaffeinate had inserted
  ```
  adjustedLength = Math.max(protocols.length, 1)
  ```
  to be safe, but we know that there has to always be at least one
  protocol, so we don't have to worry about protocols.length being 0 here.

- Refactor Connection#getState

  The decaffeinate translation of this method was not very clear, so we've
  rewritten it to be more natural.

- Simplify destructure assignment in connection.js

- Remove unnecessary use of Array.from from action_cable.js.erb

- Refactor ActionCable#createConsumer and #getConfig

  This addresses these two decaffeinate cleanup suggestions:
  - DS104: Avoid inline assignments
  - DS207: Consider shorter variations of null checks

- Remove unnecessary code created because of implicit returns in action_cable.js.erb

  This removes the `return` statements that were returning the value of
  console.log and those from methods that just set and unset the
  `debugging` flag.

- Remove decaffeinate suggestion about avoiding top-level this

  In this case, the top-level `this` is intentional, so it's okay to
  ignore this suggestion.

- Remove decaffeinate suggestions about removing unnecessary returns

  I did remove some of the return statements in previous commits, where
  it seemed appropriate. However, the rest of these should probably remain
  because the return values have been exposed through the public API. If
  we want to break that contract, we can do so, but I think it should be
  done deliberately as part of a breaking-API change (separate from this
  coffeescript -> ES2015 conversion)

- Remove unused `unsupportedProtocol` variable from connection.js

  Leaving this would cause eslint to fail

- Refactor Subscriptions methods to avoid `for` ... `of` syntax

  Babel transpiles `for` ... `of` syntax to use `Symbol.iterator`, which
  would require a polyfill in applications that support older browsers.

  The `for` ... `of` syntax was produced by running `decaffeinate`, but in
  these instances a simpler `map` should be sufficient and avoid any
  `Symbol` issues.

rmacklin committed:
Convert ActionCable javascript to ES2015 modules with modern build environment

We've replaced the sprockets `//= require` directives with ES2015
imports. As a result, the ActionCable javascript can now be compiled
with rollup (like ActiveStorage already is).

- Rename action_cable/index.js.erb -> action_cable/index.js

- Add rake task to generate a javascript module of the ActionCable::INTERNAL ruby hash

  This will allow us to get rid of ERB from the actioncable javascript,
  since it is only used to interpolate ActionCable::INTERNAL.to_json.

- Import INTERNAL directly in ActionCable Connection module

  This is necessary to remove a load-order dependency conflict in the
  rollup-compiled build. Using ActionCable.INTERNAL would result in a
  runtime error:
  ```
  TypeError: Cannot read property 'INTERNAL' of undefined
  ```
  because ActionCable.INTERNAL is not set before the Connection module
  is executed.

  All other ActionCable.* references are executed inside of the body of a
  function, so there is no load-order dependency there.

- Add eslint and eslint-plugin-import devDependencies to actioncable

  These will be used to add a linting setup to actioncable like the one
  in activestorage.

- Add .eslintrc to actioncable

  This lint configuration was copied from activestorage

- Add lint script to actioncable

  This is the same as the lint script in activestorage

- Add babel-core, babel-plugin-external-helpers, and babel-preset-env devDependencies to actioncable

  These will be used to add ES2015 transpilation support to actioncable
  like we have in activestorage.

- Add .babelrc to actioncable

  This configuration was copied from activestorage

- Enable loose mode in ActionCable's babel config

  This generates a smaller bundle when compiled

- Add rollup devDependencies to actioncable

  These will be used to add a modern build pipeline to actioncable like
  the one in activestorage.

- Add rollup config to actioncable

  This is essentially the same as the rollup config from activestorage

- Add prebuild and build scripts to actioncable package

  These scripts were copied from activestorage

- Invoke code generation task as part of actioncable's prebuild script

  This will guarantee that the action_cable/internal.js module is
  available at build time (which is important, because two other modules
  now depend on it).

- Update actioncable package to reference the rollup-compiled files

  Now that we have a fully functional rollup pipeline in actioncable, we
  can use the compiled output in our npm package.

- Remove build section from ActionCable blade config

  Now that rollup is responsible for building ActionCable, we can remove
  that responsibility from Blade.

- Remove assets:compile and assets:verify tasks from ActionCable

  Now that we've added a compiled ActionCable bundle to version control,
  we don't need to compile and verify it at publish-time.

  (We're following the pattern set in ActiveStorage.)

- Include compiled ActionCable javascript bundle in published gem

  This is necessary to maintain support for depending on the ActionCable
  javascript through the Sprockets asset pipeline.

- Add compiled ActionCable bundle to version control

  This mirrors what we do in ActiveStorage, and allows ActionCable to
  continue to be consumed via the sprockets-based asset pipeline when
  using a git source instead of a published version of the gem.

javan commented:
Exemplary pull request, @rmacklin! I will give this a closer look soon.

javan commented:
> I have some additional questions and ideas that we could discuss to improve this.

I'm all ears :)

rmacklin commented:
> Exemplary pull request, @rmacklin! I will give this a closer look soon.

Thank you! 🍻

> > I have some additional questions and ideas that we could discuss to improve this.
> 
> I'm all ears :)

Great! Here goes:

1. I'm not familiar with [Blade](https://github.com/javan/blade) (although I love the command `bundle exec blade runner` 😄), but it appears that's what was previously responsible for building the ActionCable javascript bundle. Now that this PR has added rollup to handle the build, is it correct to assume we can/should remove this: https://github.com/rails/rails/blob/8df7ed3b88fc9f19446d7207a745a331893b81cd/actioncable/blade.yml#L9-L14 ?

2. On a related note, it looks like Blade is currently used to run some javascript tests in https://github.com/rails/rails/tree/8df7ed3b88fc9f19446d7207a745a331893b81cd/actioncable/test/javascript . I ran `bundle exec rake test:integration` and visited `http://0.0.0.0:9876/` and still got [34 assertions of 34 passed, 0 failed](https://user-images.githubusercontent.com/1863540/46835159-3a18e800-cd62-11e8-8c5c-f31f886d1db6.png) when using the rollup-compiled bundle. So it doesn't seem like a change is necessary, but I'm still wondering if we _want_ to continue using Blade as the javascript test runner? Or do we want to move to a node-based test framework (e.g. `mocha`) now that were using a node-based build tool (`rollup`)?
   - I took a peak at activestorage to see what the pattern was there, but I didn’t see any javascript tests in https://github.com/rails/rails/tree/8df7ed3b88fc9f19446d7207a745a331893b81cd/activestorage/test (I could have missed something though).

3. There are some circular dependencies in the ActionCable modules (they're present in the CoffeeScript version on `master`, too, but they're surfaced more visibly in this PR now that we're using ES2015 module syntax). In particular:
   - `connection.js` imports `index.js` to use `ActionCable.log` and `ActionCable.WebSocket`
   - `connection_monitor.js` imports `index.js` to use `ActionCable.log`

   These circular dependencies don't seem to break anything, but what are your thoughts about refactoring to avoid them? I think we could break the circle by adding separate `config.js` and `logger.js` modules.
   - `startDebugging`, `stopDebugging` and `log` would be moved to the `logger.js` module. If desired, we could keep these methods on the main `ActionCable` export by delegating to `logger`, but I think it's cleaner/simpler to skip that. It would be a small breaking API change, but could easily be patched in applications that didn't want to update their usages of these methods:
     ```js
     import ActionCable from 'actioncable'; // stays the same between actioncable v5.x and v6

     // this patch would keep the remaining code working without modification
     Object.defineProperties(ActionCable, {
       startDebugging: {
         get() { return ActionCable.Logger.startDebugging; },
         set(value) { ActionCable.Logger.startDebugging = value; }
       },
       stopDebugging: {
         get() { return ActionCable.Logger.stopDebugging; },
         set(value) { ActionCable.Logger.stopDebugging = value; }
       },
       log: {
         get() { return ActionCable.Logger.log; },
         set(value) { ActionCable.Logger.log = value; }
       }
     });

     ActionCable.startDebugging();
     ActionCable.log('whatever');
     ```
   - Configuration of `ActionCable.WebSocket` and `ActionCable.logger` would be moved to the `config.js` module. Again, I think it would be okay to make this a small breaking API change, especially because I would assume the vast majority of `ActionCable` users do not override the default configuration. Users that _do_ override the defaults _and_ don't want to update their code could still patch the old API into their applications:
     ```js
     Object.defineProperties(ActionCable, {
       WebSocket: {
         get() { return ActionCable.Config.WebSocket; },
         set(value) { ActionCable.Config.WebSocket = value; }
       },
       logger: {
         get() { return ActionCable.Config.logger; },
         set(value) { ActionCable.Config.logger = value; }
       }
     });
     ```

Okay, I believe that was everything I wanted to discuss. I think (2) and (3) could be addressed in follow-up PRs if we decide they are worth addressing. Let me know if my assumption is correct re: (1) and I can push a new commit to this PR. And please also let me know if you have any comments on the existing changes in this PR!

javan commented:
> **1** … Now that this PR has added rollup to handle the build, is it correct to assume we can/should remove this

Yeah, you can remove those lines from `.blade.yml`. Doing so means we're assuming the compiled `app/assets/javascripts/action_cable.js` will be present so we may want to add it to version control [like Active Storage does](https://github.com/rails/rails/blob/99c87ad2474d5c5b6e52ceac34c3cf9f9cb57f9f/activestorage/app/assets/javascripts/activestorage.js).

> **2** …wondering if we *want* to continue using Blade as the javascript test runner?

Ultimately, no. I'd like to drop Blade entirely and switch to something like Karma. Let's leave that for a separate change though if the tests are running/passing now.

> **3** There are some circular dependencies in the ActionCable modules (they're present in the CoffeeScript version on master, too, but they're surfaced more visibly in this PR now that we're using ES2015 module syntax)

They're technically not circular in the CoffeeScript version since it doesn't have any true modules to `require` / `import`.

> I think we could break the circle by adding separate `adapters.js` and `logger.js` modules. … If desired, we could keep these methods on the main `ActionCable` export by delegating to `logger`, but I think it's cleaner/simpler to skip that. It would be a small breaking API change, but could easily be patched in applications that didn't want to update their usages of these methods

I think you can preserve the current API while still adding separate modules and eliminating the circular dependencies. Something like: 

```js
// logger.js
export default {
  logger: window.console,
  enabled: false,

  log: (...messages) {
    if (this.enabled) {
      messages.push(Date.now())
      this.logger.log("[ActionCable]", ...messages)
    }
  }
}

// index.js
import logger from "./logger"

const ActionCable = {
  // …

  set logger(value) {
    logger.logger = value,
  }
  
  startDebugging() {
    logger.enabled = true
  },

  stopDebugging() {
    logger.enabled = false
  },

  log(...messages) {
    logger.log(...messages)
  }
}
```

rmacklin commented:
> Ultimately, no. I'd like to drop Blade entirely and switch to something like Karma. Let's leave that for a separate change though if the tests are running/passing now.

Sounds good

> Yeah, you can remove those lines from `.blade.yml`. Doing so means we're assuming the compiled `app/assets/javascripts/action_cable.js` will be present so we may want to add it to version control [like Active Storage does](https://github.com/rails/rails/blob/99c87ad2474d5c5b6e52ceac34c3cf9f9cb57f9f/activestorage/app/assets/javascripts/activestorage.js).

👍 I added the compiled bundle to source control and removed that section of `blade.yml`.

Consequently, I removed the related `assets:compile` and `assets:verify` rake tasks. Now that the compiled bundle is in source control, verifying the bundle can happen before changes are merged, rather than at publish-time. We could potentially add those verifications to the test suite, but I haven't done that in this PR for a couple of reasons:

- For one, I didn't see these verifications in ActiveStorage, which I suspect is because we are trusting that the rollup will generate a proper UMD module when we specify `format: "umd"` in its config.
- Additionally, if we plan to switch to something like Karma for the javascript tests, that might be a better time to write the test that verifies that the module can be required in node. (And we could take it further than only verifying that ActionCable can be required - we could also decide to test ActionCable's functionality in a node environment.)

In the meantime, we should be able to catch breakages to the module format during code review, since git will now track the diff of the compiled bundle. That said, let me know if you think we absolutely need to replace the `assets:verify` checks with something else as part of this PR.

rmacklin commented:
> > There are some circular dependencies in the ActionCable modules (they're present in the CoffeeScript version on `master`, too, but they're surfaced more visibly in this PR now that we're using ES2015 module syntax).
>
> They're technically not circular in the CoffeeScript version since it doesn't have any true modules to `require` / `import`.

Yeah that's true; I just meant that conceptually the structure of these files and their dependencies wasn't updated by the switch to ES2015 modules.

It's just that before there was an implicit dependency from `ActionCable.Connection` to `ActionCable.log` at runtime, whereas in the ES2015 module version, the dependencies are explicit via the static `import` statements, which leads to the warning about circular dependencies at compile time. (It's just a warning though, and it still compiles correctly. Anyway, it sounds like you're in favor of breaking the circular dependencies, so I went ahead with that.) 

> I think you can preserve the current API while still adding separate modules and eliminating the circular dependencies.

Yeah, you definitely can preserve the API. That's what I meant by:

> If desired, we could keep these methods on the main `ActionCable` export by delegating to `logger`, but I think it's cleaner/simpler to skip that.

The reason I think it's cleaner to skip that is because I think it's better for a package to provide one interface. The reason being: if my application only uses `log` via an export from the `actioncable/logger` module, I shouldn't even ship the code for the unused `ActionCable.log` alias to my end users. (If this were server code, it wouldn't really matter, but any additional javascript that we have to ship to the client comes at a cost paid by the every end user.)

That said, I can understand the desire to avoid breaking backwards compatibility so that users can upgrade more easily. I just pushed some new changes that keep the API backwards compatible for the compiled bundle but avoid shipping the old API for users that consume the untranspiled source code.

That seems like the best of both worlds: upgrading won't break anything, but if you want to import from `actioncable/src` you'll need to update any usages of `log` and configuration of the WebSocket adapter (but if you aren't using `log` or changing the adapter, you wouldn't need to change anything). And if you use a module-aware/tree-shaking bundler, your application will not include any functions that you don't import, such as `startDebugging` and `stopDebugging` (so your bundle will be even smaller). With these changes, [initial example bundle savings that I reported](https://github.com/rails/rails/pull/34177#issue-221694016) improve to 3.4K (30.4%).

rmacklin commented on actioncable/app/javascript/action_cable/subscriptions.js:
```diff
@@ -0,0 +1,90 @@
+
+  reject(identifier) {
+    const result = []
+    for (let subscription of this.findAll(identifier)) {
```
I'm going to refactor this and the `notify` method below to avoid the use of `for` ... `of` syntax. Babel transpiles `for` ... `of` syntax to use `Symbol.iterator`, which would require a polyfill if your application supports browsers that do not implement `Symbol.iterator`. Running `decaffeinate` produced the `for` ... `of` syntax but in these instances a simpler `forEach` should be sufficient and avoid any `Symbol` issues.

javan commented on actioncable/app/javascript/action_cable/index_with_api_bridge.js:
```diff
@@ -0,0 +1,30 @@
+ActionCable.startDebugging = startDebugging
+ActionCable.stopDebugging = stopDebugging
+
+export default ActionCable
```
I think this should be implemented in `index.js` on the `ActionCable` object directly so this PR is purely about the conversion from CoffeeScript.

rmacklin replied:
Ok, I will move the module changes that are not related to the conversion from CoffeeScript into a separate PR.

I was planning to rebase this branch off of the latest `master` anyway, so when I do that I will completely remove those commits from this PR (rather than reverting their changes in a new commit).

javan commented on actioncable/app/javascript/action_cable/connection.js:
```diff
@@ -0,0 +1,157 @@
+    return this.isState("open", "connecting")
+  }
+
+  // Private
```
Add a `\n` here, please

rmacklin replied:
Added in https://github.com/rmacklin/rails/commit/12615fe6637a7374df7a0f1da0f87e6df91a811d - will squash during the rebase.

javan commented on actioncable/app/javascript/action_cable/subscriptions.js:
```diff
@@ -0,0 +1,90 @@
+    }
+
+    const result = []
+    subscriptions.forEach((subscription) => {
```
Ditto for returning a `.map()` result

rmacklin replied:
Changed to `map` in https://github.com/rmacklin/rails/commit/42136bf67276803d66fc2084e8b668d2071926ae - will squash during the rebase.

rmacklin commented on actioncable/app/javascript/action_cable/subscriptions.js:
```diff
@@ -0,0 +1,90 @@
+
+  reject(identifier) {
+    const result = []
+    this.findAll(identifier).forEach((subscription) => {
```
Changed to `map` in https://github.com/rmacklin/rails/commit/42136bf67276803d66fc2084e8b668d2071926ae - will squash during the rebase.

rmacklin commented:
Per https://github.com/rails/rails/pull/34177#discussion_r226360038, I have moved the module changes that are not related to the conversion from CoffeeScript into a separate PR: https://github.com/rails/rails/pull/34258

rmacklin commented:
CI is all good now!

staugaard commented:
Let's do this.

javan commented:
@rmacklin, sorry for the delay. We (Basecamp) have our company meetup this week up. I'll pick this back up next week.

rmacklin commented:
Thanks for the update! Looking forward to getting this wrapped up :)

javan commented on actioncable/app/javascript/action_cable/index.js:
```diff
@@ -0,0 +1,61 @@
+      messages.push(Date.now())
+      this.logger.log("[ActionCable]", ...messages)
+    }
+  }
```
Apologies if I wasn't clear in https://github.com/rails/rails/pull/34177#discussion_r225986157. I'm +1 on extracting a separate `logger` module to eliminate the circular dependencies. The interface here can just delegate to it, like I sketched in https://github.com/rails/rails/pull/34177#issuecomment-429460253.

rmacklin replied:
I did extract a separate logger module and eliminate the circular dependencies, but that's in https://github.com/rails/rails/pull/34258. Can you take a look at that PR? It's not exactly what you sketched but it's very similar. The main difference is that it enables applications to import exactly the ActionCable methods they need which offers the best tree shaking we can provide, while still keeping the index.js entry point clean and minimal.

javan replied:
That PR looks very nice! I don't see any reason to keep those changes separate now. Would you prefer to close this PR and work from that one?

rmacklin replied:
Great! I think I'll just pull those commits into this branch and close that PR, since this one has most of the historical discussion.

javan commented on actioncable/app/javascript/action_cable/connection.js:
```diff
@@ -0,0 +1,150 @@
+
+  getState() {
+    for (let state in WebSocket) { const value = WebSocket[state]; if (value === (this.webSocket ? this.webSocket.readyState : undefined)) { return state.toLowerCase() } }
+    return null
```
This can be rewritten more naturally now. Don't need to preserve the single-line direct CoffeeScript translation as long as the result is the same.

rmacklin replied:
Yeah I don't like the translation of this line either. Is this what you had in mind:

```js
if (this.webSocket) {
  Object.keys(WebSocket).forEach(state => {
    if (WebSocket[state] === this.webSocket.readyState) {
      return state.toLowerCase()
    }
  })
}

return null
```
or something else?

rmacklin replied:
Do you like this better:

```js
if (this.webSocket) {
  for (let state in WebSocket) {
    if (WebSocket[state] === this.webSocket.readyState) {
      return state.toLowerCase()
    }
  })
}

return null
```

I also considered some alternatives, but they used methods that would require a polyfill for browsers that don't support ES2015 natively. That would be a new requirement for ActionCable users, so I wanted to avoid it.

javan replied:
Yeah that's much better.

rmacklin replied:
Ok, I pushed that refactoring in https://github.com/rmacklin/rails/commit/dc5c137985943b7a39e99e02c78a7918e8a90833

javan commented on actioncable/app/javascript/action_cable/action_cable.js:
```diff
@@ -0,0 +1,43 @@
+  logger.enabled = false
+}
+
+export { Connection, ConnectionMonitor, Consumer, INTERNAL, log, Subscription, Subscriptions }
```
Follow up to https://github.com/rails/rails/pull/34258#discussion_r229386444:

Ha, that's a pretty bad test! It just confirms that yes in fact JavaScript does allow assigning expando properties wherever you'd like. What it doesn't test, and what I'm pretty sure won't work, is actually using the assigned adapter here: https://github.com/rails/rails/blob/629b2d6333451cf751bceb5e0e834a6e1e1c15c3/actioncable/app/javascript/action_cable/connection.js#L38

javan replied:
Unless I'm missing something? I don't see the adapter interface being exported here.

rmacklin replied:
Ah yeah, that's a bad test (but it does fail if I remove the setter defined here: https://github.com/rails/rails/blob/629b2d6333451cf751bceb5e0e834a6e1e1c15c3/actioncable/app/javascript/action_cable/index.js#L12).

Setting the adapter is implicitly tested via the `consumerTest` test cases, because `consumerTest` sets `ActionCable.WebSocket` here: https://github.com/rails/rails/blob/405aef3dede8938c8eb9789f6d6722864faddb09/actioncable/test/javascript/src/test_helpers/consumer_test_helper.coffee#L15

If I remove the setter referenced above, then the tests declared with `consumerTest` fail.

Here's a more direct test of the line you referenced (not part of the current test suite): https://github.com/rails/rails/blob/4489cd5b03a58d062ca51f21e41092d28652c7cd/actioncable/test/javascript/src/unit/connection_test.coffee#L1-L13

It passes with no changes and fails if I remove the setter referenced above.

javan replied:
Ah, my bad! I totally missed that there were separate [`index.js`](https://github.com/rails/rails/blob/dc5c137985943b7a39e99e02c78a7918e8a90833/actioncable/app/javascript/action_cable/index.js) and [`action_cable.js`](https://github.com/rails/rails/blob/dc5c137985943b7a39e99e02c78a7918e8a90833/actioncable/app/javascript/action_cable/action_cable.js) files. I would prefer they be combined so that `index.js` is the main (and only) export for the `ActionCable` interface.

sveinn replied:
I've been following this PR out of excitement for this change and I wanted to add my thoughts here. I was not too fond of the original approach that had `index.js` and a separate `index_with_api_bridge.js` which was rather messy. But I very much like the simpler `index.js` and `action_cable.js` split (IMO the latter could be renamed to just `cable.js` though).

I support keeping `index.js` and `cable.js` separate because if we combined them, we'd lose the ability for applications to tree-shake unused functions from ActionCable. I don't see why we should sacrifice that in order to have one fewer source file.

Even with `index.js` and `cable.js` as separate files, the `index.js` entry point is the main export. It has the entire `ActionCable` interface, including compatibility with the `ActionCable.WebSocket = ...` and `ActionCable.logger = ...` interface from ActionCable v5.x, and it's what you get when you simply `import ActionCable from "actioncable"` (or use `const ActionCable = require("ActionCable")`, or depend on the global variable from the UMD module). That's excellent.

But by also having the separate `cable.js` file with individually exported functions, we're giving users the ability to import `createConsumer` directly and avoid shipping everything else that's not used in their application. I think that's worth having the additional source file.

For those of us working on an application with a [performance budget](https://infrequently.org/2017/10/can-you-afford-it-real-world-web-performance-budgets/), shaving even a few KB off our dependencies gives us considerably more room to grow our application's code. This is why people like @developit and @lukeed have spent lots of time refactoring their libraries to shave off a couple hundred bytes here and there. I'm not suggesting ActionCable should worry to that extent. But I can't see why not to offer the additional tree-shaking friendly savings that come for free if we keep `index.js` and `cable.js` separate.

rmacklin replied:
I agree that we shouldn't sacrifice the tree-shaking benefits.

However, if we don't want to have the two separate source files, there is an alternative that I would be fine with: We could also export `adapters` from `action_cable.js` and move `action_cable.js` to `index.js`. Then all functionality would be available from the individual exports of the new `index.js` (former `action_cable.js`), but we would be slightly changing the public API. I don't think that's a huge cause for concern because:
- We can share the contents of what is currently `index.js` in the upgrade guide as a way for users to patch the old API into their applications when they upgrade, until they migrate to the new API.
- We could go further and ship a codemod that they can run to automatically translate their application to the new API. (If you're not familiar, this is something [react has been doing](https://github.com/reactjs/react-codemod) to make API upgrades easy for users.) I'd be happy to look into writing the codemod.

javan replied:
> I support keeping `index.js` and `cable.js` separate because if we combined them, we'd lose the ability for applications to tree-shake unused functions from ActionCable. I don't see why we should sacrifice that in order to have one fewer source file.

Can you share an example of that? How would you cherry-pick your imports for optimal tree-shaking with those separate files and what would the difference be if they were combined?

javan replied:
For this PR, I'd prefer to have everything in `index.js` so there's a single export point. That way it's a clear, straightforward record of the conversion from CoffeeScript to plain JS.

Then, if you'd like, follow up with a second PR that splits the files more optimally and documents the before/after tree-shaking improvements.

rmacklin replied:
Ok, I’ll move the changes that are not the direct CoffeeScript conversion back to https://github.com/rails/rails/pull/34258.

rmacklin replied:
Done. This PR is now only including the CoffeeScript to ES2015 translation and there's only `index.js`. Let's proceed with getting this one merged now and continue further module structure discussion at https://github.com/rails/rails/pull/34258#discussion_r229372522.

javan replied:
Thanks. FWIW, I'm +1 on your change to export individual functions like `export function createConsumer()` instead of `export const ActionCable`, and would be happy to have that included in this PR. I'd just like the `logger` and `WebSocket` interface to be similarly exported, and from the same file (`index.js`).

rmacklin replied:
Since there's no separate `action_cable.js` in this PR anymore, and we can evaluate the benefits of having that file in #34258, I'm marking this thread resolved.

rmacklin commented:
It looks like there was a broken test in ActiveRecord when this ran on CI, but it was [fixed shortly after](https://github.com/rails/rails/commit/3b2ea8b68352c1c4534ff30a086a16d00e8a12cf). I'm going to close and reopen the PR again to retrigger CI.

rmacklin commented:
CI is passing again!

Are we good to merge this and continue further discussion and improvements in subsequent PRs?

javan commented on actioncable/app/javascript/action_cable/index.js:
```diff
@@ -0,0 +1,61 @@
+ActionCable.Subscription = Subscription
+ActionCable.Subscriptions = Subscriptions
+
+export default ActionCable
```
We can save a few bytes by declaring these in the export directly.

<details><summary>View diff</summary>

```diff
--- a/actioncable/app/javascript/action_cable/index.js
+++ b/actioncable/app/javascript/action_cable/index.js
@@ -5,7 +5,13 @@ import INTERNAL from "./internal"
 import Subscription from "./subscription"
 import Subscriptions from "./subscriptions"
 
-const ActionCable = {
+export default {
+  Connection,
+  ConnectionMonitor,
+  Consumer,
+  INTERNAL,
+  Subscription,
+  Subscriptions,
   WebSocket: window.WebSocket,
   logger: window.console,
 
@@ -14,7 +20,7 @@ const ActionCable = {
       const urlConfig = this.getConfig("url")
       url = (urlConfig ? urlConfig : this.INTERNAL.default_mount_path)
     }
-    return new ActionCable.Consumer(this.createWebSocketURL(url))
+    return new Consumer(this.createWebSocketURL(url))
   },
 
   getConfig(name) {
@@ -50,12 +56,3 @@ const ActionCable = {
     }
   }
 }
-
-ActionCable.Connection = Connection
-ActionCable.ConnectionMonitor = ConnectionMonitor
-ActionCable.Consumer = Consumer
-ActionCable.INTERNAL = INTERNAL
-ActionCable.Subscription = Subscription
-ActionCable.Subscriptions = Subscriptions
-
-export default ActionCable
```

</details><br>

```
app/assets/javascripts/action_cable.js
  before: 19664 bytes
   after: 19586 bytes
```

rmacklin replied:
Done in https://github.com/rmacklin/rails/commit/ad035aac03ecaed8c784d3f8bf958914cf515c6b

javan commented on actioncable/.babelrc:
```diff
@@ -0,0 +1,8 @@
+{
+  "presets": [
+    ["env", { "modules": false } ]
```
Let's use Babel's loose mode. The resulting JS is smaller and more like what the CoffeeScript version produced.
```diff
--- a/actioncable/.babelrc
+++ b/actioncable/.babelrc
@@ -1,6 +1,6 @@
 {
   "presets": [
-    ["env", { "modules": false } ]
+    ["env", { "modules": false, "loose": true } ]
   ],
   "plugins": [
     "external-helpers"
```
Full diff: https://gist.github.com/javan/387bec1ca052e36a9904177f87fed2de

```
app/assets/javascripts/action_cable.js
  before: 19664 bytes
   after: 17900 bytes
```

javan replied:
(I plan to do the same for Active Storage too)

rmacklin replied:
Done in https://github.com/rmacklin/rails/commit/746a19c9127fbda263571faa2e2f138e93e0cd3d

javan commented:
Just left a couple more little comments. Once those are addressed can squash your commits, please? 1 to 3 meaningful commits is fine. After that we're good to merge.

rmacklin commented:
Addressed the two changes. I'm busy for the next couple of hours, but I'll rebase and squash the commits later today. In the meantime, can you take a look at my comments on https://github.com/rails/rails/pull/34258#discussion_r229825330 and let me know if https://github.com/rmacklin/rails/commit/832225ea02145a499caa68d64f8d6d1ce78c7264 is the right interpretation of what you'd like to happen there?

rmacklin commented:
> One more thing… We need to include the new `app/assets/javascripts/action_cable.js` file in the published gem for Sprockets support. Currently, [`app/` isn't the `s.files` array](https://github.com/rails/rails/blob/746a19c9127fbda263571faa2e2f138e93e0cd3d/actioncable/actioncable.gemspec#L20). [Here's `activestorage.gemspec`](https://github.com/rails/rails/blob/d80925b169f0db950697868f5659a67cc03e8654/activestorage/activestorage.gemspec#L20) for comparison.

Added in https://github.com/rmacklin/rails/commit/85fa4fc1a6997d5870b7d002a02a23f66f4945a6

I'm ready to start squashing, but is there anything else?

I've also opened https://github.com/rails/rails/pull/34370 with what I believe is the approach you were describing in the https://github.com/rails/rails/pull/34177#pullrequestreview-169889965 thread. It's simpler than the approach in https://github.com/rails/rails/pull/34258 because avoids introducing a new `action_cable.js` file at the cost of slightly modifying the API for setting adapters. I am fine with either approach, so if you like https://github.com/rails/rails/pull/34370 I can continue working there and close https://github.com/rails/rails/pull/34258, or vice versa.

rmacklin commented:
Updated with the squashed version. 

The first commit is only file moving and renaming. I kept that as a separate commit to allow the git history of the files to be preserved (i.e. ensure git will track these as file renames rather than unrelated file additions/deletions). In particular, you can easily trace the git blame back to the original authorship:

![file_history_intact](https://user-images.githubusercontent.com/1863540/47926653-5221ea00-de7e-11e8-827c-625f760e2894.gif)

The remaining meaningful commits are:
- Run automated conversions on the coffeescript files
- Manually clean up converted code (I wanted to keep a clear separation between the conversion done by machine and the subsequent cleanup done by hand)
- Convert to ES2015 modules and modify the build environment to support this (I also didn't want to muddy the waters by mixing the `decaffeinate` cleanup with the conversion of IIFEs and sprockets `//= require`s into ES2015 modules)

rmacklin commented:
CI is back in the green 🎉

javan commented:
Looks good! 

Quick question: Have you tried using the compiled `action_cable.js` in a working application (that actually makes a WebSocket connection)?

rmacklin commented:
I created the [`use-es2015-actioncable` branch](https://github.com/rmacklin/anycable_demo/tree/use-es2015-actioncable) in this repository: https://github.com/rmacklin/anycable_demo

(It's a fork of https://github.com/anycable/anycable_demo, but the repository supports regular ActionCable in addition to AnyCable)

Running:
```
git clone git@github.com:rmacklin/anycable_demo.git
cd anycable_demo
git checkout use-es2015-actioncable
bin/setup
bin/rails server
```
and then visiting `localhost:3000` demonstrates a working application using this PR's version of `action_cable.js` to make a real WebSocket connection.

javan commented:
Thank you!

rmacklin commented:
Thanks for your help!
